src/Npgsql/NpgsqlDataSource.cs

```diff
@@ -51,6 +67,11 @@ public abstract class NpgsqlDataSource : DbDataSource
 
     readonly ILogger _connectionLogger;
 
+    /// <summary>
+    /// Semaphore to ensure we don't perform type loading and mapping setup concurrently for this data source.
+    /// </summary>
+    readonly SemaphoreSlim _setupMappingsSemaphore = new(1);
```

Comment:
Shouldn't we dispose it while disposing data source?

Reply:
And another thought, maybe make `NpgsqlDataSource` resilient against multiple disposes?

Reply:
> Shouldn't we dispose it while disposing data source?

Good catch - will dispose the SemaphoreSlim in NpgsqlDataSource.Dispose. Note that this slightly is relevant for @Brar's work in #4722 (/cc @NinoFloris) - if ReloadTypes and Clear/ClearAll race, ReloadTypes could throw ObjectDisposedException (as it tries to take the semaphore just as its being disposed). This is in theory something we don't want, but I think it's acceptable for ReloadTypes specifically.

Reply:
> And another thought, maybe make NpgsqlDataSource resilient against multiple disposes?

Good point, adding this (please take a look).